File: 2024/haskell/day06a/Main.hs
module Main ( main ) where

import Data.Array ( Array )
import qualified Data.Array as Array
import Data.Bifunctor ( bimap )
import Data.Set ( Set )
import qualified Data.Set as Set
import System.Environment ( getArgs, getProgName )
import System.Exit ( exitFailure )

type Position = (Int, Int)

data Lab = Lab { start :: Position
               , direction :: Int
               , grid :: Array (Int, Int) Char
               , numRows :: Int
               , numCols :: Int
               }

usage :: IO ()
usage = do
  progname <- getProgName
  putStrLn $ "usage: " ++ progname ++ " <file>"
  exitFailure

parse :: String -> Lab
parse contents =
    let rows = lines contents
        numRows = length rows
        numCols = length (head rows)
        start = head $ [(r, c) | r <- [0..(numRows - 1)], c <- [0..(numCols - 1)], rows !! r !! c == '^']
        elems = [((i, j), c) | i <- [0..numRows - 1], j <- [0..numCols - 1], let c = (rows !! i) !! j]
        grid = Array.array ((0, 0), (numRows - 1, numCols - 1)) elems
    in Lab { start = start, direction = 0, grid = grid, numRows = numRows, numCols = numCols }

directions :: [(Int, Int)]
directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

inBounds :: Lab -> Position -> Bool
inBounds lab pos@(r, c) = r >= 0 && r < numRows lab && c >= 0 && c < numCols lab

guardPositions :: Lab -> [(Position, Int)]
guardPositions lab =
    let pos = start lab
        dir = direction lab
        guardPositions' pos dir positions =
            let positions' = positions ++ [(pos, dir)]
                nextPos = bimap (fst pos +) (snd pos +) (directions !! dir)
            in if not (inBounds lab nextPos)
                then positions'
                else let dir' = (if grid lab Array.! nextPos  == '#'
                                    then (dir + 1) `mod` 4
                                    else dir)
                         pos' = bimap (fst pos +) (snd pos +) (directions !! dir')
                     in guardPositions' pos' dir' positions'
    in guardPositions' pos dir []

process :: String -> Int
process contents =
    let lab = parse contents
        moves = guardPositions lab
    in Set.size $ Set.fromList (map fst moves)

main :: IO ()
main = do
  args <- getArgs
  case args of
    [filename] -> do
      contents <- readFile filename
      let result = process contents
      putStrLn ("result = " ++ show result)
    _ -> usage

File: 2024/rust/day06a/src/main.rs
use std::collections::HashSet;
use std::env;
use std::fs;
use std::process;
use std::time::SystemTime;

type Position = (usize, usize);

struct Lab {
    start: Position,
    direction: usize,
    grid: Vec<Vec<char>>,
    num_rows: usize,
    num_cols: usize,
}

fn usage() {
    let progname = env::args().next().unwrap();
    println!("usage: {progname} <file>");
    process::exit(1);
}

fn parse(contents: &str) -> Lab {
    let mut grid: Vec<Vec<char>> = vec![];
    for line in contents.lines() {
        grid.push(line.chars().collect::<Vec<char>>());
    }
    let num_rows: usize = grid.len();
    let num_cols: usize = grid[0].len();
    let mut start: Position = (0, 0);
    for (row, row_chars) in grid.iter_mut().enumerate() {
        for (col, ch) in row_chars.iter_mut().enumerate() {
            if *ch == '^' {
                start = (row, col);
                *ch = '.';
            }
        }
    }
    Lab {
        start,
        direction: 0,
        grid,
        num_rows,
        num_cols,
    }
}

fn guard_positions(lab: &Lab) -> Vec<(Position, usize)> {
    let directions = [(-1, 0), (0, 1), (1, 0), (0, -1)];
    let mut positions: Vec<(Position, usize)> = vec![];
    let mut pos = lab.start;
    let mut dir = lab.direction;
    loop {
        positions.push((pos, dir));
        let next_pos = (
            pos.0 as isize + directions[dir].0,
            pos.1 as isize + directions[dir].1,
        );
        if next_pos.0 < 0
            || next_pos.0 >= lab.num_rows as isize
            || next_pos.1 < 0
            || next_pos.1 >= lab.num_cols as isize
        {
            return positions;
        }
        if lab.grid[next_pos.0 as usize][next_pos.1 as usize] == '#' {
            dir = (dir + 1) % 4;
        }
        pos = (
            (pos.0 as isize + directions[dir].0) as usize,
            (pos.1 as isize + directions[dir].1) as usize,
        );
    }
}

fn process(contents: &str) -> u64 {
    let lab = parse(contents);
    let positions = guard_positions(&lab);
    let unique_positions: HashSet<Position> = positions
        .into_iter()
        .map(|p| p.0)
        .collect::<HashSet<Position>>();
    unique_positions.len() as u64
}

fn calc_runtime(start_time: &SystemTime) -> String {
    let mut elapsed = start_time.elapsed().unwrap().as_micros() as f64;
    let mut units = "μs";
    if elapsed > 1000.0 {
        elapsed /= 1000.0;
        units = "ms";
    }
    if elapsed > 1000.0 {
        elapsed /= 1000.0;
        units = "s";
    }
    format!("(elapsed time: {elapsed} {units})")
}

fn main() {
    let start_time = SystemTime::now();
    if env::args().count() < 2 {
        usage();
    }
    let filename = env::args().nth(1).unwrap();
    let contents = fs::read_to_string(filename).expect("read of input file failed");
    let result = process(&contents);
    let duration = calc_runtime(&start_time);
    println!("result = {result}  {duration}");
}
